### PR TITLE
Vg 521 wd backend

### PR DESCRIPTION
**Pull request by CedricCabessa**
Draft for WD backend
See commits message for more information

### CODE REVIEW

File: pb/keychain/service.proto [hunk 1]
```diff
@@ -110,6 +110,8 @@ message CreateKeychainRequest {
   Scheme scheme = 3;
   uint32 lookahead_size = 4;
   ChainParams chain_params = 5;
+  uint32 index = 6;
```
Comment (onyb) [inline pb/keychain/service.proto:113]: Let's call this `account_index`.

Also for future, we may want to stop having the account index in `FromChainCode` message. Let's not do it now, else it'll break the current Lama code.

File: pb/keychain/service.proto [hunk 2]
```diff
@@ -110,6 +110,8 @@ message CreateKeychainRequest {
...
+  string info = 7;
```
Comment (onyb) [inline pb/keychain/service.proto:114]: Please add some comments about this field, and what it's useful for. Also, let's specify the schema expected by this field for the known use-cases we have. Remember that this file is the only source of information for external teams.

Also, what do you think of `metadata` as a name? It indicates that it is a non-essential field.

File: pkg/keystore/wd.go [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+
+	if _, err := rdb.Ping(context.Background()).Result(); err != nil {
+		return nil, fmt.Errorf("Pinging redis failed: %w", err)
+	}
+
+	return &WDKeystore{
+		db:     rdb,
+		client: bitcoin.NewBitcoinClient(),
+	}, nil
+}
```
Comment (onyb) [inline pkg/keystore/wd.go:34]:
```suggestion
type WDKeystore struct {
	baseRedisKeystore
}

func NewWDKeystore(redisOpts *redis.Options) (*WDKeystore, error) {
	rdb := redis.NewClient(redisOpts)

	if _, err := rdb.Ping(context.Background()).Result(); err != nil {
		return nil, fmt.Errorf("Pinging redis failed: %w", err)
	}

	baseKeystore := baseRedisKeystore{
		db:     rdb,
		client: bitcoin.NewBitcoinClient(),
	}

	return &WDKeystore{baseKeystore }, nil
}
```

Then you should be able to add the following to `redis_common.go`.

```go
type baseRedisKeystore struct {
	db     *redis.Client
	client bitcoin.CoinServiceClient
}

func (s *baseRedisKeystore) Get(id uuid.UUID) (KeychainInfo, error) {
	var meta Meta

	err := get(s.db, id.String(), &meta)
	if err != nil {
		return KeychainInfo{}, ErrKeychainNotFound
	}

	return meta.Main, nil
}
```

The `client` and `db` fields on `WDKeystore` will work as usual.

This pattern gives you the benefits of inheritance but using composition. You can read about it [here](https://travix.io/type-embedding-in-go-ba40dd4264df).

**Review by onyb — CHANGES_REQUESTED** [review]

File: pkg/keystore/wd.go (hunk 3 above)
Comment (CedricCabessa) [reply pkg/keystore/wd.go:34]: very cool, thx !

File: pb/keychain/service.proto (hunk 1 above)
Comment (CedricCabessa) [reply pb/keychain/service.proto:113]: >  Let's not do it now, else it'll break the current Lama code.

What do you want me to do :sweat_smile: ?
I've discussed with Edouard, he is ok to pass the index. Also this information is mandatory to speak "wd"
Are you saying to remove "account_index" later on?

**Comment by CedricCabessa** [comment]
I'll squash the commit before merge

File: pb/keychain/service.proto (hunk 1 above)
Comment (onyb) [reply pb/keychain/service.proto]: I was saying that `FromChainCode` also has the `account_index` which is duplication. If we pass it in `CreateKeychainRequest`, then we don't need to provide it in `FromChainCode`.

**Review by onyb — APPROVED** [review]

**Pull request merged**